### DISCUSSION TITLE
Why doesn't `AuthenticateAsync` set context's User property?

### DISCUSSION THREAD

**Discussion by voroninp** [dotnet/aspnetcore#51695] [Q&A]
In `PolicyEvaluator` class principal is built from [all specified schemes](https://github.com/dotnet/aspnetcore/blob/73e0b536d411a38c745b2a7ee5b04f04ee4fe6ea/src/Security/Authorization/Policy/src/PolicyEvaluator.cs#L40).

Whereas direct call to [`HttpContext.AuthenticateAsync`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.authenticationhttpcontextextensions.authenticateasync?view=aspnetcore-8.0#microsoft-aspnetcore-authentication-authenticationhttpcontextextensions-authenticateasync(microsoft-aspnetcore-http-httpcontext-system-string)) has no effect. Authentication middleware [sets `User` only for the default scheme](https://github.com/dotnet/aspnetcore/blob/73e0b536d411a38c745b2a7ee5b04f04ee4fe6ea/src/Security/Authentication/Core/src/AuthenticationMiddleware.cs#L65).

Should not it be the responsibility of `AuthenticationService`, and should not it be the same as in policy evaluator? 
I'd expect, that every call to `Authenticate` extends (or replaces for the same scheme?) context's user with the new principal same way `PolicyEvaluator` does.

**Comment by voroninp** [comment]
@davidfowl, your answer would be much appreciated ;-)

**Reply by davidfowl** [reply]
I don’t think that call should have a side effect that results in the user property being set.  I absolutely dislike that the policy evaluator does that

**Reply by voroninp** [reply]
But where should it be set then?

**Reply by davidfowl** [reply]
The caller of that method should set it if they need it to be set.

**Reply by voroninp** [reply]
Hm... but with auth policies/authorize attribute that's going to be some code in framework anyway, isn't it?

**Reply by davidfowl** [reply]
This method is a low level method that returns a result. It shouldn’t have a side effect of changing the current user. That feels very broken. 

It can be called from anywhere in the request pipeline to get the authentication result (for example SignalR calls it to get the result of authentication).

**Reply by voroninp** [reply]
Ah, ok I got you. I agree `Authenticate`'s signature implies it's a query rather than a command.
